Bluetooth: Fix callback handling in ECC Public Key generation

michal-narajowski opened the pull request:
Commit d6c34c4644c8fa771ed138b166be4500867c9332 changed the behavior
slightly but didn't update the documentation. The callback will now be
reset to NULL once the key is generated. Calling bt_pub_key_gen()
multiple times before the key is finished would result in creation of an
infinite loop. This could happen when an application calls mesh_init()
and mesh_reset() in quick succession.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

sjanc commented on subsys/bluetooth/host/ecc.h:
```diff
@@ -23,9 +23,8 @@ struct bt_pub_key_cb {
- *  key has been generated, and will be used to notify of new generation
- *  processes (NULL as key).
+ *  Generate a new ECC Public Key. The callback will persist until the key
+ *  has been generated. Subsequent calls will overwrite the stored callback.
```
I'd change this to something like:
Generate a new ECC Public Key. Provided cb must persists until callback is called. Function with same cb parameter shall not be called again until callback is called.

sjanc reviewed (changes requested):
This should be fixed in prov.c as bt_mesh_prov_reset_state() is calling this API in invalid way (if multiple callbacks are neeed there should be either multiple instances of struct bt_pub_key_cb provided or internal list of custom callbacks should be implemented in prov.c

trond-snekvik commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -9330,7 +9330,6 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 		return -ENOTSUP;
 	}
 
-	new_cb->_next = pub_key_cb;
```
This removes the linked list property of the callbacks, and makes it impossible to have multiple pending callbacks for the pubkey generation. As far as I know, we'd like to be able to call pubkey gen from multiple modules and still get the callbacks in each of them once it's ready.

If we really don't want to have multiple callbacks, we can just pass the function pointer here instead, and get rid of the for loop in `le_pkey_complete`, but I think we need it. I think a better solution would be to scan through the list or use the `_next` pointer to determine whether it's already in the list. Perhaps use sys/slist.h in this case.

trond-snekvik replied:
@joerchan pointed out this issue: #23292.
It's currently not possible to freeze or otherwise ensure sole control of the ECC keypair in Zephyr, mostly because of limitations in HCI. Joakim's second proposal matches the changes in this PR, where only a single module can have its callback listed at a time. If we want to go in this direction, we should probably return `-EBUSY` if there's already a callback present, so some other module doesn't come in and steal the callback away from us.

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -3856,7 +3856,7 @@ static void le_pkey_complete(struct net_buf *buf)
 	}
 
-	for (cb = pub_key_cb; cb; cb = cb->_next) {
+	for (cb = pub_key_cb; cb && cb->func; cb = cb->_next) {
```
When does the cb->func become NULL, and should the for loop really stop because of this?
Maybe that should just be in if inside the for loop to skip calling the func in that case?

Commit:
Bluetooth: Fix callback handling in ECC Public Key generation

Commit d6c34c4644c8fa771ed138b166be4500867c9332 changed the behavior
slightly but didn't update the documentation. The callback will now be
reset to NULL once the key is generated. Calling bt_pub_key_gen()
multiple times before the key is finished would result in creation of an
infinite loop. This could happen when an application calls mesh_init()
and mesh_reset() in quick succession. Clarify the behavior of the API in
the documentation.

Also passing a NULL argument would result in an undefined behavior, so
add a check to match the behavior described in documentation.

Signed-off-by: Michał Narajowski <michal.narajowski@codecoup.pl>

michal-narajowski commented:
@joerchan @trond-snekvik I switched to using `sys/slist.h` API and added a check if the callback is already registered.

trond-snekvik commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -64,8 +65,18 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
+		SYS_SLIST_FOR_EACH_CONTAINER(&pub_key_cb_slist, cb, node) {
+			if (cb == new_cb) {
+				BT_WARN("Callback already registered");
+				new_cb = NULL;
```
return `-EALREADY`?

trond-snekvik commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -40,6 +40,7 @@ bool bt_pub_key_is_debug(uint8_t *pub_key)
 
 int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 {
+	struct bt_pub_key_cb *cb;
```
Can be moved inside the if-scope on line 68

trond-snekvik commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -77,7 +88,7 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 		BT_ERR("Sending LE P256 Public Key command failed");
 		atomic_clear_bit(bt_dev.flags, BT_DEV_PUB_KEY_BUSY);
-		pub_key_cb = NULL;
+		sys_slist_init(&pub_key_cb_slist);
```
Should probably call each callback with a NULL pointer to avoid dropping callbacks silently

sjanc commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -180,11 +199,13 @@ void bt_hci_evt_le_pkey_complete(struct net_buf *buf)
 	}
 
-	pub_key_cb = NULL;
+	sys_slist_init(&pub_key_cb_slist);
```
aren't callbacks suppose to be preserved?    (calling bt_pub_key_gen(NULL) is allowed according to docs to refresh keys)

joerchan replied:
Maybe update the documentation to match the new behavior then? There are only users of this API, as long as both work with the current behavior that should be fine.

sjanc replied:
yeap, lets keep this as is and update docs

carlescufi commented:
@sjanc please re-review

sjanc commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -64,8 +64,18 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
 
-	new_cb->_next = pub_key_cb;
-	pub_key_cb = new_cb;
+	if (new_cb) {
```
this suggest that passing NULL new_cb is valid, but since list is cleared after key is generated it makes no sense to call with NULL.  I'd suggest to return error if callback is NULL and remove note about updating key in such case from bt_pub_key_gen() documentation.

sjanc commented on subsys/bluetooth/host/ecc.c:
```diff
@@ -64,8 +64,18 @@ int bt_pub_key_gen(struct bt_pub_key_cb *new_cb)
+
+		SYS_SLIST_FOR_EACH_CONTAINER(&pub_key_cb_slist, cb, node) {
+			if (cb == new_cb) {
+				BT_WARN("Callback already registered");
```
This means that once cannot call bt_pub_key_gen() directly from result callback.  I'm fine with this as long as it is documented.
